apps/data-pipeline/sample-program-scraper/src/index.ts

```diff
@@ -0,0 +1,715 @@
+      const href = $(a).attr("href");
+      if (href?.startsWith("/")) {
+        programPaths.push(href);
+      }
+    });
+  }
+  return Array.from(new Set(programPaths));
+}
+
+/*
+  Extracts a unique ID from the catalogue URL
+  Example: https://catalogue.uci.edu/.../computerscience_bs/ => "computerscience_bs"
+ */
+function generateProgramId(url: string): string {
+  // Regex pulls the final path segment from the catalogue URL (e.g. /foo/bar/ -> bar).
+  const match = url.match(/\/([^/]+)\/?$/);
+  if (!match) {
+    logger.warn(`Could not extract ID from URL: ${url}`);
+    return "";
+  }
+  return match[1];
+}
+
+/*
+  Transforms the original sample program format to have a Fall, Winter, Spring structure
+  @param sampleYears The original sample years data from the scraper
+  @returns The transformed data with Fall, Winter, Spring structure
+*/
+function transformToTermStructure(
+  sampleYears: ScrapedSampleYear[],
+): UnvalidatedSampleProgramEntry[] {
+  const transformedProgram: UnvalidatedSampleProgramEntry[] = [];
+
+  for (const yearData of sampleYears) {
+    const fall: string[] = [];
+    const winter: string[] = [];
+    const spring: string[] = [];
+
+    const curriculum = yearData.curriculum;
+
+    for (const row of curriculum) {
+      if (row.length >= 1 && row[0].trim()) {
+        fall.push(row[0].trim());
+      }
+      if (row.length >= 2 && row[1].trim()) {
+        winter.push(row[1].trim());
+      }
+      if (row.length >= 3 && row[2].trim()) {
+        spring.push(row[2].trim());
+      }
+    }
+
+    transformedProgram.push({
+      year: yearData.year,
+      fall,
+      winter,
+      spring,
+    });
+  }
+
+  return transformedProgram;
+}
+
+// Parse course code into department and course number for functionality lookup
+function parseCourseCode(courseCode: string): { dept: string; num: string } | null {
+  // Match pattern: letters/spaces/& followed by numbers and optional letters
+  const match = courseCode.match(/^([A-Z0-9&\s]+?)(\d+[A-Z]*)$/);
+
+  if (!match) return null;
+
+  return {
+    dept: match[1].trim(),
+    num: match[2],
+  };
+}
+
+async function lookupCourseId(
+  db: ReturnType<typeof database>,
+  courseCode: string,
+): Promise<string | null> {
+  try {
+    const parsed = parseCourseCode(courseCode);
+
+    if (!parsed) {
+      return null;
+    }
+
+    const [foundCourse] = await db
+      .select({ id: course.id })
+      .from(course)
+      .where(and(eq(course.department, parsed.dept), eq(course.courseNumber, parsed.num)))
+      .limit(1);
+
+    return foundCourse?.id ?? null;
+  } catch (error) {
+    logger.warn(`Failed to lookup course: ${courseCode}`, error);
+    return null;
+  }
+}
+
+/*
+  Transforms course code strings to course IDs by looking them up in the database.
+  Falls back to original string if course not found.
+ */
+async function transformCourseCodesToIds(
+  db: ReturnType<typeof database>,
+  scrapedPrograms: {
```

Comment:
Whoa! These inline types are a little unwieldy. I think they would look better outlined, either just above this function or with the other types at the top of the file.